fableous-fe/src/components/Canvas.tsx

```diff
@@ -18,417 +18,609 @@ import { ControllerRole, ToolMode, WSMessage, WSMessageType } from "../Data";
+        const shape = textShapes[editingText];
+        const { key } = event;
+        if (key === "Escape" || key === "Enter") {
+          setEditingText(0);
+        }
+        if (key.length === 1 || key === "Backspace") {
+          if (key.length === 1) {
+            shape.text += key;
+          } else if (key === "Backspace") {
+            if (shape.text)
+              shape.text = shape.text.substring(0, shape.text.length - 1);
+          }
+          placeText(
+            (shape.x1 + shape.x2) / 2,
+            (shape.y1 + shape.y2) / 2,
+            editingText,
+            shape.text,
+            shape.fontSize
+          );
+        }
+      },
+      [editingText, placeText, textShapes]
+    );
 
-  useEffect(() => {
-    const canvas = canvasRef.current;
-    canvas.width = canvas.offsetWidth * SCALE;
-    canvas.height = canvas.offsetWidth * ASPECT_RATIO * SCALE;
-    setAllowDrawing(role !== ControllerRole.Hub);
-    switch (role) {
-      case ControllerRole.Story:
-        setToolMode(ToolMode.Text);
-        break;
-      case ControllerRole.Character:
-        setToolMode(ToolMode.Paint);
-        break;
-      case ControllerRole.Background:
-        setToolMode(ToolMode.Paint);
-        break;
-      default:
-        setToolMode(ToolMode.None);
-    }
-    const ws = wsRef.current;
-    ws?.addEventListener("message", readMessage);
-    return () => ws?.removeEventListener("message", readMessage);
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [role, wsRef]);
+    // setup on component mount
+    useEffect(() => {
+      const canvas = canvasRef.current;
+      const ctx = canvas.getContext("2d");
+      canvas.width = canvas.offsetWidth * SCALE;
+      canvas.height = canvas.offsetWidth * ASPECT_RATIO * SCALE;
+      setAllowDrawing(role !== ControllerRole.Hub);
+      switch (role) {
+        case ControllerRole.Story:
+          setToolMode(ToolMode.Text);
+          break;
+        case ControllerRole.Character:
+          setToolMode(ToolMode.Paint);
+          break;
+        case ControllerRole.Background:
+          setToolMode(ToolMode.Paint);
+          break;
+        default:
+          setToolMode(ToolMode.None);
+      }
+      if (ctx) {
+        ctx.textAlign = "center";
+        ctx.textBaseline = "middle";
+      }
+      const ws = wsRef.current;
+      ws?.addEventListener("message", readMessage);
+      return () => {
+        ws?.removeEventListener("message", readMessage);
+      };
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [role, wsRef]);
+
+    // initialize keyboard listener
+    useEffect(() => {
+      if (layer === ControllerRole.Story) {
+        window.addEventListener("keydown", onKeyDown);
+      }
+      return () => {
+        window.removeEventListener("keydown", onKeyDown);
+      };
+    }, [layer, onKeyDown]);
+
+    // start text layer animation
+    useEffect(() => {
+      window.requestAnimationFrame(refreshText);
```

Comment:
wanted to confirm, just for own understanding, so refreshText will draw texts to canvas, and this will trigger whenever textShapes changes, as thats in dep list of refreshText's useCallback()?

Reply:
that was obsolete as it used to be triggered via JS intervals, however it was rather choppy and buggy. now, refreshText() will be trigerred via the browser's (window) animation frame event triggers, resulting in a much more consistent 60 fps update